File: Day 23 - Binary Tree/Maximum path sum.java
class Solution {
    int ans = Integer.MIN_VALUE;
    
    public int solve(TreeNode node) {
        if (node == null) return 0;
        int l = Math.max(solve(node.left), 0);
        int r = Math.max(solve(node.right), 0);
        int cur = node.val + l + r;
        ans = Math.max(ans, cur);
        return node.val + Math.max(l, r);
    }
    
    public int maxPathSum(TreeNode root) {
        solve(root);
        return ans;
    }
}

File: Day 23 - Binary Tree/Maximum path sum.cpp
class Solution
{
public:
  int ans = INT_MIN;
  int solve(TreeNode *node)
  {
    if (!node)
      return 0;
    int l = max(solve(node->left), 0);
    int r = max(solve(node->right), 0);
    int cur = node->val + l + r;
    ans = max(ans, cur);
    return node->val + max(l, r);
  }
  int maxPathSum(TreeNode *root)
  {
    solve(root);
    return ans;
  }
};